### DISCUSSION TITLE
Covering index as a primary key

### DISCUSSION THREAD

**Discussion by epandurski** [sqlalchemy/sqlalchemy#5866] [Usage Questions] [answered]
**Describe your question**

I use PostgreSQL. I want the index created for the primary key of a table to include some other columns (a covering index).

In SQLAlchemy 1.4.0b1, can I define the primary key index to "cover" additional columns? I am using `DeclarativeBase` table schema style.

I tried to pass "postgresql_include" argument to `PrimaryKeyConstraint`, and it did not work. I also tried to declare the primary key in `__mapper_args__`, and define the primary key index separately with `CreateIndex` adding the "postgresql_include" argument, but it did not work too.

Will this use case be supported in SQLAlchemy 1.4?

 

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**

**Complete stack trace, if applicable**

**Versions**
 - OS: Linux
 - Python: 3.7
 - SQLAlchemy: 1.4.0b1
 - Database: PostgreSQL
 - DBAPI: psycopg2

**Additional context**

**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
hi there -

can you help me out a bit and please describe an example of the CREATE TABLE statement you are looking for ?   thanks!

**Comment by epandurski** [comment]
I would like to generate some SQL like this:
```sql
CREATE TABLE t1 (
   id: int NOT NULL,
   included_col: int NOT NULL
);
CREATE UNIQUE INDEX idx_t1_pk (id) INCLUDE (included_col);

```

**Comment by zzzeek** [comment]
So that's not a primary key constraint, that's a unique index which is achieved using the ``Index()`` construct that offers the postgresql_include parameter.

your exact SQL (with correct syntax for the CREATE TABLE) can be created with the 1.4 series as:
```

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Index
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import Table

m = MetaData()

t = Table(
    "t1", m,
    Column("id", Integer, nullable=False),
    Column("included_col", Integer, nullable=False)
)

Index("idx_t1_pk", t.c.id, postgresql_include=["included_col"], unique=True)

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
t.drop(e, checkfirst=True)
t.create(e)
```

output:

```
CREATE TABLE t1 (
	id INTEGER NOT NULL, 
	included_col INTEGER NOT NULL
)


2021-01-23 11:39:50,749 INFO sqlalchemy.engine.Engine [no key 0.00005s] {}
2021-01-23 11:39:50,752 INFO sqlalchemy.engine.Engine CREATE UNIQUE INDEX idx_t1_pk ON t1 (id) INCLUDE (included_col)

```

**Comment by epandurski** [comment]
I am using the ORM, and the unit of work pattern, and I want the ORM to know what is the primary key. The problem is that I can not both:
1) Generate the proper CREATE TABLE/ CREATE INDEX statements
2) Tell the ORM what the primary key for this table is.

 Could you show me how I can use declarative base to achieve this. And be able to use the table with the session?

**Comment by zzzeek** [comment]
it depends on if you want there to also be a primary key constraint on the table or not.

the most straightforward way is just a normal declarative setup, then add the index as well:

```
class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    included_col = Column(Integer)

Index("idx_t1_pk", A.id, postgresql_include=["included_col"], unique=True)

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.create_all(e)

```

if you want the table to not actually include a primary key constraint, which would be unusual, you can tell the mapper the primary key columns directly using [primary_key](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.mapper.params.primary_key) like this:

```
class A(Base):
    __tablename__ = 'a'

    id = Column(Integer)
    included_col = Column(Integer)

    __mapper_args__ = {"primary_key": [id]}

Index("idx_t1_pk", A.id, postgresql_include=["included_col"], unique=True)

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

```

the above table wont use PostgreSQL SERIAL though and the "id" wont be autoincrementing.

**Comment by zzzeek** [comment]
also found [this link](https://stackoverflow.com/a/11370534/34549) regarding this topic.

**Comment by epandurski** [comment]
> > it depends on if you want there to also be a primary key constraint on the table or not.
> > the most straightforward way is just a normal declarative setup, then add the index as well:
> > ```
> > class A(Base):
> >     __tablename__ = 'a'
> > 
> >     id = Column(Integer, primary_key=True)
> >     included_col = Column(Integer)
> > 
> > Index("idx_t1_pk", A.id, postgresql_include=["included_col"], unique=True)
> > 
> > e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
> > Base.metadata.create_all(e)
> > ```
> > 

I tried this, and it seems to work, but this seems to create 2 redundant indexes: One for the `primary_key=True`, which does not include the "included_col" and one named "idx_t1_pk", which includes the "included_col". I want to have only one index which includes the "included_col".

> > 
> > if you want the table to not actually include a primary key constraint, which would be unusual, you can tell the mapper the primary key columns directly using [primary_key](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.mapper.params.primary_key) like this:
> > ```
> > class A(Base):
> >     __tablename__ = 'a'
> > 
> >     id = Column(Integer)
> >     included_col = Column(Integer)
> > 
> >     __mapper_args__ = {"primary_key": [id]}
> > 
> > Index("idx_t1_pk", A.id, postgresql_include=["included_col"], unique=True)
> > 
> > e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
> > Base.metadata.drop_all(e)
> > Base.metadata.create_all(e)
> > ```
> > 

I tried this as well, and it does not seem to work. Here is the error that I get (`debtor_id` is the PK in the table `debtor`, and `status_flags` is the included column):

```
db_session = <sqlalchemy.orm.scoping.scoped_session object at 0x7f7180682e48>

    @pytest.fixture(scope='function')
    def debtor(db_session):
        debtor = m.Debtor(debtor_id=123, status_flags=0)
        debtor.activate()
        db_session.add(debtor)
>       db_session.commit()

tests/test_routes.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<string>:2: in commit
    ???
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:1283: in commit
    self._transaction.commit(_to_root=self.future)
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:709: in commit
    self._prepare_impl()
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:688: in _prepare_impl
    self.session.flush()
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3099: in flush
    self._flush(objects)
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3239: in _flush
    transaction.rollback(_capture_exception=True)
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py:71: in __exit__
    with_traceback=exc_tb,
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/util/compat.py:180: in raise_
    raise exception
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/session.py:3199: in _flush
    flush_context.execute()
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py:442: in execute
    rec.execute(self)
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py:612: in execute
    uow,
../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py:246: in save_obj
    insert,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

base_mapper = <Mapper at 0x7f7183bc63c8; Debtor>, uowtransaction = <sqlalchemy.orm.unitofwork.UOWTransaction object at 0x7f7180366390>
mapper = <Mapper at 0x7f7183bc63c8; Debtor>
table = Table('debtor', MetaData(), Column('debtor_id', BigInteger(), table=<debtor>, nullable=False), Column('status_flags', ...Column('config_latest_update_id', BigInteger(), table=<debtor>, nullable=False, default=ColumnDefault(1)), schema=None)
insert = <generator object _collect_insert_commands at 0x7f7180641318>, bookkeeping = True

    def _emit_insert_statements(
        base_mapper,
        uowtransaction,
        mapper,
        table,
        insert,
        bookkeeping=True,
    ):
        """Emit INSERT statements corresponding to value lists collected
        by _collect_insert_commands()."""
    
        cached_stmt = base_mapper._memo(("insert", table), table.insert)
    
        execution_options = {"compiled_cache": base_mapper._compiled_cache}
    
        for (
            (connection, pkeys, hasvalue, has_all_pks, has_all_defaults),
            records,
        ) in groupby(
            insert,
            lambda rec: (
                rec[4],  # connection
                set(rec[2]),  # parameter keys
                bool(rec[5]),  # whether we have "value" parameters
                rec[6],
                rec[7],
            ),
        ):
    
            statement = cached_stmt
    
            if (
                not bookkeeping
                or (
                    has_all_defaults
                    or not base_mapper.eager_defaults
                    or not connection.dialect.implicit_returning
                )
                and has_all_pks
                and not hasvalue
            ):
                # the "we don't need newly generated values back" section.
                # here we have all the PKs, all the defaults or we don't want
                # to fetch them, or the dialect doesn't support RETURNING at all
                # so we have to post-fetch / use lastrowid anyway.
                records = list(records)
                multiparams = [rec[2] for rec in records]
    
                c = connection._execute_20(
                    statement, multiparams, execution_options=execution_options
                )
    
                if bookkeeping:
                    for (
                        (
                            state,
                            state_dict,
                            params,
                            mapper_rec,
                            conn,
                            value_params,
                            has_all_pks,
                            has_all_defaults,
                        ),
                        last_inserted_params,
                    ) in zip(records, c.context.compiled_parameters):
                        if state:
                            _postfetch(
                                mapper_rec,
                                uowtransaction,
                                table,
                                state,
                                state_dict,
                                c,
                                last_inserted_params,
                                value_params,
                                False,
                                c.returned_defaults
                                if not c.context.executemany
                                else None,
                            )
                        else:
                            _postfetch_bulk_save(mapper_rec, state_dict, table)
    
            else:
                # here, we need defaults and/or pk values back.
    
                records = list(records)
                if (
                    not hasvalue
                    and connection.dialect.insert_executemany_returning
                    and len(records) > 1
                ):
                    do_executemany = True
                else:
                    do_executemany = False
    
                if not has_all_defaults and base_mapper.eager_defaults:
                    statement = statement.return_defaults()
                elif mapper.version_id_col is not None:
                    statement = statement.return_defaults(mapper.version_id_col)
                elif do_executemany:
                    statement = statement.return_defaults(*table.primary_key)
    
                if do_executemany:
                    multiparams = [rec[2] for rec in records]
    
                    c = connection._execute_20(
                        statement, multiparams, execution_options=execution_options
                    )
    
                    if bookkeeping:
                        for (
                            (
                                state,
                                state_dict,
                                params,
                                mapper_rec,
                                conn,
                                value_params,
                                has_all_pks,
                                has_all_defaults,
                            ),
                            last_inserted_params,
                            inserted_primary_key,
                            returned_defaults,
                        ) in util.zip_longest(
                            records,
                            c.context.compiled_parameters,
                            c.inserted_primary_key_rows,
                            c.returned_defaults_rows or (),
                        ):
                            for pk, col in zip(
                                inserted_primary_key,
                                mapper._pks_by_table[table],
                            ):
                                prop = mapper_rec._columntoproperty[col]
                                if state_dict.get(prop.key) is None:
                                    state_dict[prop.key] = pk
    
                            if state:
                                _postfetch(
                                    mapper_rec,
                                    uowtransaction,
                                    table,
                                    state,
                                    state_dict,
                                    c,
                                    last_inserted_params,
                                    value_params,
                                    False,
                                    returned_defaults,
                                )
                            else:
                                _postfetch_bulk_save(mapper_rec, state_dict, table)
                else:
                    for (
                        state,
                        state_dict,
                        params,
                        mapper_rec,
                        connection,
                        value_params,
                        has_all_pks,
                        has_all_defaults,
                    ) in records:
                        if value_params:
                            result = connection._execute_20(
                                statement.values(value_params),
                                params,
                                execution_options=execution_options,
                            )
                        else:
                            result = connection._execute_20(
                                statement,
                                params,
                                execution_options=execution_options,
                            )
    
                        primary_key = result.inserted_primary_key
>                       assert primary_key
E                       AssertionError

../../.virtualenvs/swpt_debtors/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py:1225: AssertionError
================================================================== short test summary info ===================================================================
ERROR tests/test_procedures.py::test_initiate_running_transfer - AssertionError
ERROR tests/test_routes.py::test_initiate_running_transfer - AssertionError

```

It looks like SQLAlchemy thinks that my table does not have a primary key. I believe this is a regression, because the same thing works with SQLAlchemy 1.3.x (that is: without the "postgresql_include", which is not supported). Here is the schema that I am trying to generate: https://github.com/epandurski/swpt_debtors/blob/master/swpt_debtors/models.py . All tests pass with SQLalchemy 1.3.x, and most of the tests fail with 1.4.0b1. Let me know if I can give you more information to better identify the problem. (Or to figure out my mistake.)

> > 
> > the above table wont use PostgreSQL SERIAL though and the "id" wont be autoincrementing.

Another probably unrelated problem is that the "alembic" version that I use does not recognize the `postgresql_include=['status_flags']` argument.

**Reply by zzzeek** [reply]
> I tried this, and it seems to work, but this seems to create 2 redundant indexes: One for the `primary_key=True`, which does not include the "included_col" and one named "idx_t1_pk", which includes the "included_col". I want to have only one index which includes the "included_col".

That's Postgresql which always creates an index to correspond to a primary key constraint.    Googling doesnt seem to find it but I would wonder if there's a way to either alter this index, or replace it, with one that has covering columns.  the stackoverflow answer seems to be talking about Postgresql before index-only scans were inroduced.   


> 
> I tried this as well, and it does not seem to work. Here is the error that I get (`debtor_id` is the PK in the table `debtor`, and `status_flags` is the included column):
> 
> ```
> db_session = <sqlalchemy.orm.scoping.scoped_session object at 0x7f7180682e48>
> 
>     @pytest.fixture(scope='function')
>     def debtor(db_session):
>         debtor = m.Debtor(debtor_id=123, status_flags=0)
>         debtor.activate()
>         db_session.add(debtor)
> >       db_session.commit()

I've identified that assertion as indeed causing a regression for various primary key insert schemes where the table itself does not have a PrimaryKeyConstraint on it, although it's usually schemes that make use of Sequence or Identity to generate the primary key value.   Your example above where you pass the PK value explicitly does not as easily lead down that codepath unless I add some other defaults on the mapping.   #5867 has a bunch of combinations that lead to that erroneous assertion.

**Reply by epandurski** [reply]
Thanks for your help, and all the great work that you are doing.

**Comment by gordthompson** [comment, marked as answer]
Resolved.
